### PR TITLE
Add const to TextInputFormatter

### LINKED ISSUE #114401: Add const constructor to TextInputFormatter
## Use case

Currently, form controls are const constructible. But there is single thing preventing it to const construct. It is TextInputFormatter. 

## Proposal

Add const constructor to `TextInputFormatter`
```
abstract class TextInputFormatter {
  const TextInputFormatter();

  // rest of codes
}
```
This makes subclasses of `TextInputFormatter` are const constructible.

### PR DESCRIPTION
**Pull request by shogohida**
Fixes #114401 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by shogohida** [comment]
It seems test failures are not related to the change I made so I decided to open this pull request

**Comment by LongCatIsLooong** [comment]
Thanks for contributing. The change makes sense to me, but the test failures are related, "Linux analyze" is complaining there's missing documentation. Also this seems to be testable, could you add a test for that?

**Comment by shogohida** [comment]
@LongCatIsLooong 
Thanks for your comment! I have 2 questions.

1. I couldn't see that "Linux analyze" is complaining there's missing documentation... I saw the error message below but nothing else. Could you show me the specific point where Linux analyze is complaining? I will fix the code according to that point.

```
For your convenience, the error messages reported above are repeated here:
Command exited with exit code 1 but expected: zero exit code.
Command: bin/flutter analyze --flutter-repo
Relative working directory: .
Command exited with exit code 1 but expected: zero exit code.
Command: bin/flutter analyze --flutter-repo --watch --benchmark
Relative working directory: .
```

2. Is there any good example in Flutter code I can refer to in order to write a test? I wasn't sure if I should write a test to examine const constructer can be created in TextInputFormatter class or its sub classes.

Thanks for your support!

**Comment by LongCatIsLooong** [comment]
You should be able to find plenty of examples here: https://github.com/flutter/flutter/blob/master/packages/flutter/test/services/text_formatter_test.dart.

The test failure is here: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8798526700594975537/+/u/Analyze/analyze/test_stdout. The new public constructor needs to be documented.

**Comment by shogohida** [comment]
@LongCatIsLooong 
I added documentation and tests but it seems tests I added are wrong.... I tried to create tests referring to the file you shared but I don't know yet what kind of tests I should add.

Could you give me a clear code example of tests? It would be so helpful. Thanks in advance.

File: packages/flutter/test/services/text_formatter_test.dart [hunk 1]
```diff
@@ -7,6 +7,13 @@ import 'package:flutter/services.dart';
 import 'package:flutter_test/flutter_test.dart';
 
 void main() {
+  test('test const constructor', () {
+    const testValue1 = TextInputFormatter();
```
Comment (goderbauer) [inline packages/flutter/test/services/text_formatter_test.dart:11]: TextInputFormatter is an abstract class and cannot be instantiated.

Comment (shogohida) [reply packages/flutter/test/services/text_formatter_test.dart:11]: @goderbauer 
Thanks for your comment. Should I write tests for const constructors of subclasses instead?

**Comment by LongCatIsLooong** [comment]
For tests I think the ones you added were fine, just have to make sure you can add `const` to subclasses. But I don't think it makes sense to add a new const constructor that takes no parameters for every subclass.

**Comment by shogohida** [comment]
@LongCatIsLooong 
Thanks for your comment! Are these changes fine then?

https://github.com/flutter/flutter/pull/114592/commits/ad66a5d504988528aa4052d99de119795339466b

File: packages/flutter/test/services/text_formatter_test.dart [hunk 2]
```diff
@@ -7,6 +7,19 @@ import 'package:flutter/services.dart';
 import 'package:flutter_test/flutter_test.dart';
 
 void main() {
+  test('test const constructor', () {
+    const testValue1 = _SimpleTextInputFormatter();
```
[the remark this answers is not in the document]
Comment (shogohida) [reply packages/flutter/test/services/text_formatter_test.dart]: Sorry for replying late. Is my understanding correct?

https://github.com/flutter/flutter/pull/114592/commits/8b96bf068d8c3011e4b17b70ca1a8ef31801a1c4

**Comment by goderbauer** [comment]
(triage): @shogohida looks like the checks failed due to some infra issue. Could you merge in the latest master to kick the checks again? Thanks.

**Comment by shogohida** [comment]
@goderbauer 
Thanks for your comment. I tried to rebase but it seems I made mistakes in the process. I will make a new pull request from the latest master and close this one for a moment.

**Pull request closed without merging**